internal/controller/argocdcommitstatus_controller.go

```diff
@@ -0,0 +1,401 @@
+		} else if repo != application.Spec.SourceHydrator.DrySource.RepoURL {
+			return map[string][]*aggregate{}, fmt.Errorf("all applications must have the same repo configured")
+		}
+
+		aggregateItem := &aggregate{
+			application: &application,
+		}
+
+		phase := promoterv1alpha1.CommitPhasePending
+		if application.Status.Health.Status == argocd.HealthStatusHealthy && application.Status.Sync.Status == argocd.SyncStatusCodeSynced {
+			phase = promoterv1alpha1.CommitPhaseSuccess
+		} else if application.Status.Health.Status == argocd.HealthStatusDegraded {
+			phase = promoterv1alpha1.CommitPhaseFailure
+		}
+
+		// This is an in memory version of the desired CommitStatus for a single application, this will be used to figure out
+		// the aggregated phase of all applications for a particular environment
+		aggregateItem.commitStatus = &promoterv1alpha1.CommitStatus{
+			Spec: promoterv1alpha1.CommitStatusSpec{
+				Sha:   application.Status.Sync.Revision,
+				Phase: phase,
+			},
+		}
+		argoCDCommitStatus.Status.ApplicationsSelected = append(argoCDCommitStatus.Status.ApplicationsSelected, promoterv1alpha1.ApplicationsSelected{
+			Namespace:          application.GetNamespace(),
+			Name:               application.GetName(),
+			Phase:              phase,
+			Sha:                application.Status.Sync.Revision,
+			LastTransitionTime: application.Status.Health.LastTransitionTime,
+		})
+
+		aggregates[application.Spec.SourceHydrator.SyncSource.TargetBranch] = append(aggregates[application.Spec.SourceHydrator.SyncSource.TargetBranch], aggregateItem)
+	}
+
+	return aggregates, nil
+}
+
+func (r *ArgoCDCommitStatusReconciler) calculateAggregatedPhaseAndDescription(appsInEnvironment []*aggregate, resolvedSha string, mostRecentLastTransitionTime *metav1.Time) (promoterv1alpha1.CommitStatusPhase, string) {
+	var desc string
+	resolvedPhase := promoterv1alpha1.CommitPhasePending
+	pending := 0
+	healthy := 0
+	degraded := 0
+	for _, s := range appsInEnvironment {
+		if s.commitStatus.Spec.Sha != resolvedSha {
+			pending++
+		} else if s.commitStatus.Spec.Phase == promoterv1alpha1.CommitPhaseSuccess {
+			healthy++
+		} else if s.commitStatus.Spec.Phase == promoterv1alpha1.CommitPhaseFailure {
+			degraded++
+		} else {
+			pending++
+		}
+	}
+
+	// Resolve state
+	if healthy == len(appsInEnvironment) {
+		resolvedPhase = promoterv1alpha1.CommitPhaseSuccess
+		desc = fmt.Sprintf("%d/%d apps are healthy", healthy, len(appsInEnvironment))
+	} else if degraded == len(appsInEnvironment) {
+		resolvedPhase = promoterv1alpha1.CommitPhaseFailure
+		desc = fmt.Sprintf("%d/%d apps are degraded", degraded, len(appsInEnvironment))
+	} else {
+		desc = fmt.Sprintf("Waiting for apps to be healthy (%d healthy, %d degraded, %d pending)", healthy, degraded, pending)
+	}
+
+	if mostRecentLastTransitionTime != nil && time.Since(mostRecentLastTransitionTime.Time) < 5*time.Second {
+		return promoterv1alpha1.CommitPhasePending, desc
+	}
```

Comment:
```suggestion
	// Don't consider the aggregate status healthy until 5s after the most recent transition.
	// This helps avoid prematurely accepting a transitive healthy state.
	if mostRecentLastTransitionTime != nil && time.Since(mostRecentLastTransitionTime.Time) < 5*time.Second {
		return promoterv1alpha1.CommitPhasePending, desc
	}
```